Sources/AWSLambdaRuntimeCore/Documentation.docc/Proposals/0001-v2-api.md

```diff
@@ -0,0 +1,667 @@
+    func write(_ buffer: ByteBuffer) async throws
+    /// End the response stream
+    consuming func finish()
+    /// Write a response part into the stream and end the response stream
+    consuming func writeAndFinish(_ buffer: ByteBuffer) async throws
+    /// Report an error in the response stream
+    consuming func reportError(_ error: any Error) async throws
+}
+```
+
+### LambdaContext
+
+`LambdaContext` will be largely unchanged, but the `eventLoop` property as well as the `allocator` property (of type
+`ByteBufferAllocator`) will be removed.
+
+A new function `addBackgroundTask(_:)` will also be added. This will allow tasks to be run in the background while and after
+the response is/has been sent. Please note that `LambdaContext` will not be `Sendable` anymore, because in order
+to allow adding background tasks in a structured way, the `LambdaContext` will need to be backed with a `TaskGroup`.
+Since `TaskGroup` is not `Sendable` we cannot mark `LambdaContext` as `Sendable`. If users need any property out of the
+`LambdaContext` they can just extract the property and then pass it around or close over it as all the properties are
+`Sendable`.
+
+```swift
+/// A context object passed as part of an invocation in LambdaHandler handle functions.
+public struct LambdaContext {
+    /// The request ID, which identifies the request that triggered the function invocation.
+    public var requestID: String { get }
+
+    /// The AWS X-Ray tracing header.
+    public var traceID: String { get }
+
+    /// The ARN of the Lambda function, version, or alias that's specified in the invocation.
+    public var invokedFunctionARN: String { get }
+
+    /// The timestamp that the function times out.
+    public var deadline: DispatchWallTime { get }
+
+    /// For invocations from the AWS Mobile SDK, data about the Amazon Cognito identity provider.
+    public var cognitoIdentity: String? { get }
+
+    /// For invocations from the AWS Mobile SDK, data about the client application and device.
+    public var clientContext: String? { get }
+
+    /// `Logger` to log with.
+    ///
+    /// - note: The `LogLevel` can be configured using the `LOG_LEVEL` environment variable.
+    public var logger: Logger { get }
+
+    /// Schedule background work in lambda. This work must be completed before
+    /// the runtime can ask for further events. Note that AWS will continue to
+    /// charge you after the response has been returned, but work is still
+    /// processing in the background.
+    public func addBackgroundTask(_ body: sending @escaping () async -> ())
+}
+```
+
+### StreamingLambdaHandler
+
+The new `StreamingLambdaHandler` protocol is the base protocol to implement a Lambda function. Most users will not use
+this protocol and instead use the `LambdaHandler` protocol defined in the **Codable Support** section.
+
+```swift
+/// The base StreamingLambdaHandler protocol
+public protocol StreamingLambdaHandler: ~Copyable {
```

Comment:
I think making this `~Copyable` will lead to a load of issues. Can we have a motivation on why we allow non-copyable types to implement this? If we allow non-copyable types here, then anything accepting `StreamingLambdaHandler `s needs to support this which sounds like a tall ask.

Reply:
> Can we have a motivation on why we allow non-copyable types to implement this?

It is not necessary to have a non copyable Handler. We just allow it. Tbh. the motivation here is mostly, because we can.

> If we allow non-copyable types here, then anything accepting `StreamingLambdaHandler`s needs to support this which sounds like a tall ask.

I think the only thing that should accept this protocol is `LambdaRuntime` and we have support for non copyable already implemented in our prototype. What else would consume a `StreamingLambdaHandler`?

Reply:
> What else would consume a StreamingLambdaHandler?

The user's tests?

> we have support for non copyable already implemented in our prototype.

Sure, but it could cause problems further down the line, no? Like if you figure out that you will need to copy it

Reply:
> The user's tests?

I think we would provide a `TestRuntime` that does that for the user.

 > Sure, but it could cause problems further down the line, no? Like if you figure out that you will need to copy it

Makes sense. But in this case requirements would have changed so much that we should investigate a v3.

Reply:
Really not sure about that. I don't think it gives you any benefit but it restricts your code's evolution. Don't see the point :).